File: sdk/catena.py
from __future__ import annotations

import json
import urllib.request
from dataclasses import dataclass
from typing import Any


@dataclass
class CatenaClient:
    base_url: str
    api_key: str = ""

    def __post_init__(self) -> None:
        self.base_url = self.base_url.rstrip("/")

    def query(self, sql: str, params: list[Any] | None = None) -> dict[str, Any]:
        return self._post("/query", {"sql": sql, "params": params or []})

    def transaction(self, statements: list[dict[str, Any]]) -> list[dict[str, Any]]:
        payload = self._post("/transaction", {"statements": statements})
        return payload["results"]

    def _post(self, path: str, payload: dict[str, Any]) -> dict[str, Any]:
        body = json.dumps(payload).encode("utf-8")
        headers = {"Content-Type": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"
        req = urllib.request.Request(
            f"{self.base_url}{path}",
            data=body,
            headers=headers,
            method="POST",
        )
        try:
            with urllib.request.urlopen(req, timeout=30) as res:
                return json.loads(res.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            details = exc.read().decode("utf-8")
            raise RuntimeError(f"Catena request failed: {details}") from exc

File: sdk/catena.js
export class CatenaClient {
  constructor(baseURL, options = {}) {
    this.baseURL = baseURL.replace(/\/$/, "");
    this.apiKey = options.apiKey || "";
  }

  headers() {
    const headers = { "Content-Type": "application/json" };
    if (this.apiKey) {
      headers.Authorization = `Bearer ${this.apiKey}`;
    }
    return headers;
  }

  async query(sql, params = []) {
    const response = await fetch(`${this.baseURL}/query`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({ sql, params }),
    });
    const payload = await response.json();
    if (!response.ok) {
      throw new Error(payload.message || payload.error || "Catena query failed");
    }
    return payload;
  }

  async transaction(statements) {
    const response = await fetch(`${this.baseURL}/transaction`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({ statements }),
    });
    const payload = await response.json();
    if (!response.ok) {
      throw new Error(payload.message || "Catena transaction failed");
    }
    return payload.results;
  }

  async backup() {
    const response = await fetch(`${this.baseURL}/backup`, {
      method: "POST",
      headers: this.headers(),
    });
    const payload = await response.json();
    if (!response.ok) {
      throw new Error(payload.message || "Catena backup failed");
    }
    return payload;
  }

  async metrics() {
    const response = await fetch(`${this.baseURL}/metrics`, {
      headers: this.headers(),
    });
    const payload = await response.json();
    if (!response.ok) {
      throw new Error(payload.message || "Catena metrics request failed");
    }
    return payload;
  }

  async export() {
    const response = await fetch(`${this.baseURL}/export`, {
      headers: this.headers(),
    });
    if (!response.ok) {
      const payload = await response.json();
      throw new Error(payload.message || "Catena export failed");
    }
    return response.blob();
  }

  subscribe(table, onEvent) {
    const url = new URL(this.baseURL);
    url.protocol = url.protocol === "https:" ? "wss:" : "ws:";
    url.pathname = "/ws";
    if (this.apiKey) {
      url.searchParams.set("token", this.apiKey);
    }

    const socket = new WebSocket(url.toString());
    socket.addEventListener("open", () => {
      socket.send(JSON.stringify({ type: "subscribe", table }));
    });
    socket.addEventListener("message", (event) => {
      onEvent(JSON.parse(event.data));
    });
    return socket;
  }
}
